corehq/apps/cloudcare/views.py

```diff
@@ -109,21 +113,16 @@ def fetch_app(self, domain, app_id):
         return _fetch_build(domain, self.request.couch_user.username, app_id)
 
     def get_web_apps_available_to_user(self, domain, user):
-        app_access = get_application_access_for_domain(domain)
+        def is_web_app(app):
+            return app.get('cloudcare_enabled') or self.preview
+
+        apps = []
         app_ids = get_app_ids_in_domain(domain)
+        for app_id in app_ids:
+            app = self.fetch_app(domain, app_id)
+            if app and is_web_app(app) and can_user_access_web_app(app, user, domain):
+                apps.append(_format_app_doc(app))
 
-        apps = list(map(
-            lambda app_id: self.fetch_app(domain, app_id),
-            app_ids,
-        ))
-        apps = filter(None, apps)
-        apps = filter(lambda app: app.get('cloudcare_enabled') or self.preview, apps)
-        # Backwards-compatibility - mobile users haven't historically required this permission
-        if user.is_web_user() or user.can_access_any_web_apps(domain):
-            apps = filter(lambda app: user.can_access_web_app(domain, app.get('copy_of', app.get('_id'))), apps)
-        if toggles.WEB_APPS_PERMISSIONS_VIA_GROUPS.enabled(domain):
-            apps = filter(lambda app: app_access.user_can_access_app(user, app), apps)
```

Comment:
I think the new function could result in a different outcome than this if both `if` statements evaluate to true. Not sure if that is possible in practice?

Example: an app would have been eliminated from the list of apps if the app was excluded by the first `filter(...)` but would have been included in the second if it had gotten that far (effectively these two conditions are combined with `and`), whereas the new `can_user_access_web_app` function would return true because those conditions are combined with `or`.

Reply:
Yeah this was a great catch. I believe I've addressed this concern and now give precedence to permissions over groups. There are a couple of other threads in this PR that highlight similar issues. See 92ffbbc2c5c2e55fa3d713b6cc2b891c67413c02